feat(blob): Move client methods to @vercel/blob/client + renaming

vvo opened the pull request:
This pull request introduces BREAKING CHANGES. Mostly, we've been trying to separate client and server needs better. While ensuring we only export what we currently think is useful to you.

We've moved and renamed client-related utilities, including the ones to generate client tokens, to a separate entry file: `import { put, upload, handleUpload, generateClientTokenFromReadWriteToken} from '@vercel/blob/client'`.

Here are the most important BREAKING CHANGES:
- `BlobCommandOptions` type is no longer exported. This is an internal utility type. Other internal types were removed also but nobody was using them
- handleBlobUpload has moved to client/`handleUpload`
- For client (browser) uploads, please use client/`upload` instead of `put`. Also, the `handleBlobUploadUrl` option has been renamed to `handleUploadUrl`.
- `verifyCallbackSignature` is no more exported if you think you would like to use it, open an issue

---

This PR allows `clientPayload` during `upload,` so you can attach
metadata to the file uploaded to Vercel Blob. This metadata will be sent as
`tokenPayload` by default in uploading completed callbacks. And you can also enrich
it/configure it/verify it.

Usage:
```ts
// client.ts
import { upload } from '@vercel/blob/client';

upload(
  'file.txt',
  {
    access: 'public',
    handleUploadUrl: '/api/upload',
    clientPayload: JSON.stringify({ postId: 123 })
  }
);
```

```ts
const jsonResponse = await handleUpload({
  body,
  request,
  onBeforeGenerateToken: async (pathname, clientPayload) => {

    return {
       tokenPayload: clientPayload // this is the default, ideally you would verify the clientPayload (format, authorization) before generating the token, so no files are uploaded without first verifying the client claims
    };
  },
  onUploadCompleted: async ({ blob, tokenPayload }) => {
    console.log('blob upload completed', blob, tokenPayload);
  }
});
```

An example use-case of this feature would be to allow editing the images of a
blog post. To do so you would use `clientPayload: JSON.stringify({ postId: 345
})` in the browser. Then, either in onBeforeGenerateToken or in onUploadCompleted
you would verify that the client making the call is allowed to modify this
postId (authorization) and update your database accordingly.

BREAKING CHANGES:
- `metadata` in `onBeforeGenerateToken` and `onUploadCompleted` has been renamed to `tokenPayload`

vvo committed:
feat(blob): Add clientPut to clearly separate server and client uploads

- BREAKING CHANGE: renamed all API entries and documentation with "handleBlob" to "handleClient".
  If you were using handleBlobUploadUrl, handleBlobUpload, .. then replace
handleBlob* and BlobUpload* with handleClient* and ClientUpload*
- BREAKING CHANGE: `put` will no more automatically switch to client (browser)
  uploads. Use `clientPut` if you want the previous behavior
- added `clientPut` which only accept to get the token from
  handleClientUploadUrl. This method is the goto method to upload blobs from
clients (browsers, mobile..). This method do not accept parameters that are
enforced by the client token (cacheControlMaxAge, addRandomSuffix)
- renamed "browser uploads" to "client uploads" (Vercel documentation to be
  updated soon)

vvo committed:
update examples

vvo committed:
default 5 minutes edge cache

correttojs commented on packages/blob/src/index.ts:
```diff
@@ -122,6 +116,96 @@ export async function put(
+    | File,
+  options: ClientPutCommandOptions
+): Promise<PutBlobResult> {
+  if (!pathname) {
```
should we throw if `typeof window === undefined`? not 100% sure that the request comes from the client, but it could still be useful

vvo replied:
Same as https://github.com/vercel/storage/pull/347/files#r1313037483, we should not artificially limit where put() can be used only because we provide a convenience clientPut() method. Some people will not want to use it.

correttojs commented on packages/blob/src/index.ts:
```diff
@@ -76,12 +75,7 @@ export async function put(
-        pathname,
-      })
-    : getToken(options);
+  const token = getTokenFromOptionsOrEnv(options);
```
same here, could we throw in case `typeof window` is defined?

vvo replied:
I would not do so (same reasons as my comment here: https://github.com/vercel/storage/pull/347/files#r1313037483). put() can be used client or server side, the fact that we provide a wrapper (clientPut) that does the token exchange automatically should not prevent anyone to build their own client upload wrappers if they feel limited/constrained by ours (handleBlobUpload).

correttojs commented on packages/blob/src/index.ts:
```diff
@@ -122,6 +116,96 @@ export async function put(
+  handleClientUploadUrl: string;
+}
+
+export async function clientPut(
```
can we try to create a `client/index.ts` and re-export only this file?
I know we said it's fine to expose the `clientPut` function, but think that it would be harder to move it once we'll release the public beta

vvo replied:
When (= why) would we need to move some methods to /client?

As of today, I am unsure if there will be a time when we will need to move clientPut() to a different import path: all methods of the blob package can be called on the frontend. For example, if you wanted to allow list/delete/head to be called from the frontend, enforcing a specific prefix for which paths can be listed (think: per-user/per-x blob folder).

vvo replied:
I moved methods to client when they are related to client uploads in general

correttojs commented on packages/blob/README.md:
```diff
@@ -31,11 +31,30 @@ async function put(
     // and can be configured when you connect more stores to a project
     // or using Vercel Blob outside of Vercel
-    // During browser uploads, a token is automatically generated by calling the route from the `handleBlobUploadUrl` option
+    // You could also pass a client token here if you generated it yourself with `generateClientTokenFromReadWriteToken`
```
it seems a bit confusing this point, why don't we allow the client upload only from the `clientPut`? if the user wants to pass the clientToken explicitly, we can create a type like `{handleClientUploadUrl} | {clientToken}`

vvo replied:
I see `clientPut` as a convenience method companion to the handleBlobUpload (server) method. But nothing prevents someone from using the put() method directly from the frontend. Either via a regular token or a client token. There's a good probability some people might just want to roll their custom client upload doing so, and I don't want to artificially limit them doing so.

correttojs replied:
so it's possible to send a `cacheControlMaxAge` option from the client by manually creating a token. The option would be then ignored by the worker 🤔

vvo replied:
> so it's possible to send a cacheControlMaxAge option from the client by manually creating a token

Yes, by using a read write token on the client (not recommended but 🤷 ).
Yes, by creating a client token and using `put()`. In this case, this is where we could (maybe) add runtime checks and (for sure) server-side checks on what is received.

correttojs replied:
then I don't see the advantages of adding a `clientPut`, the additional typescript interface `{handleClientUploadUrl} | {cacheControlMaxAge,...}` would be enough...are there other reasons to create a `clientPut`?

vvo replied:
Yes, I think it clarifies the intent and behavior of the put method when you want to use the client upload helpers versus having a method that changes its behavior based on its arguments.

vvo committed:
Merge branch 'main' into feat/add-clientPut

vvo committed:
feat(blob): Create another entry point for all client upload needs

- BREAKING CHANGES(client uploads):
  - import { put } from '@vercel/blob' => import { upload } from '@vercel/blob/client-upload'
  - put(pathname, content, { handleBlobUploadUrl[...] }) => upload(pathname, content, { handleUploadUrl })
  - import { handleBlobUpload } from '@vercel/blob' => import { handleUpload } from '@vercel/blob/client-upload'
  - a bunch of types and helpers related to client uploads were moved (and renamed for a few of them) to '@vercel/blob/client-upload'

renamed all API entries and documentation with "handleBlob"
to "handleClient". If you were using handleBlobUploadUrl, handleBlobUpload, .. then replace handleBlob* and BlobUpload* with handleClient* and ClientUpload*

vvo committed:
Merge branch 'feat/add-clientPut' of github.com:vercel/storage into feat/add-clientPut

vvo committed:
update

vvo committed:
update

vvo committed:
update

vvo committed:
update

vvo commented on packages/blob/src/client-upload.node.test.ts:
```diff
@@ -0,0 +1,186 @@
+import type { IncomingMessage } from 'node:http';
```
I moved all index.test.ts client upload tests to this file, that's it

vvo committed:
remove useless code

vvo committed:
reduce changes

vvo committed:
change everything

vvo commented on test/next/tsconfig.json:
```diff
@@ -10,7 +10,7 @@
     "esModuleInterop": true,
     "module": "esnext",
-    "moduleResolution": "node",
+    "moduleResolution": "Node",
```
this doesn't change anything, just TS automatically set it with a capital Node somehow if you change it

vvo committed:
update

vvo committed:
update documentation

vvo committed:
Merge branch 'main' into feat/add-clientPut

vvo committed:
update

vvo committed:
update

vvo committed:
update

vvo committed:
update

vvo committed:
update

correttojs commented on packages/blob/src/client.ts:
```diff
@@ -145,30 +179,31 @@ export function getPayloadFromClientToken(
   payload: {
+    // TODO @Fabio: is this correct? I guess it should be a PutBlobResult and not HeadBlobResult unless
+    // we're doing an extra head() call somewhere?
     blob: HeadBlobResult;
```
I think at the time we implemented this feature the 2 interfaces were exactly the same
Now the event receives 
```
 const blob: BlobPutApiResult | BlobPutApiResultV0 =
    apiVersion < 2
      ? // we're faking previous response format for the vercel/blob module not to throw
        {
          url: encodeURI(`${baseUrlForBlobs}/${key}`),
          pathname,
          contentType,
          contentDisposition,
          uploadedAt: new Date().toISOString(),
          size: parseInt(request.headers.get('content-length') ?? '0', 10),
        }
      : {
          url: encodeURI(`${baseUrlForBlobs}/${key}`),
          pathname,
          contentType,
          contentDisposition,
        };
```
so `PutBlobResult` should be the correct one

vvo replied:
Updated 👍

correttojs commented on packages/blob/src/put.ts:
```diff
@@ -0,0 +1,125 @@
+  contentType: 'x-content-type',
+};
+
+export interface PutBlobResult {
```
just an idea for a separate PR...it would be nice to consume these interfaces from the worker, to ensure client and server uses the same fields

vvo replied:
Indeed, I have been reluctant to do so because then it ties both services very tightly (for example: you cannot easily add a new property to edge-functions unless you update the npm package).

correttojs commented on packages/blob/src/put.ts:
```diff
@@ -0,0 +1,125 @@
+
+type PartialBy<T, K extends keyof T> = Omit<T, K> & Partial<Pick<T, K>>;
+
+export function createPutMethod<
```
nice interface!

vvo replied:
👍

correttojs reviewed (approved):
tested the client and server upload, all good 👍
great job!

vvo committed:
update

vvo committed:
update readme

vvo committed:
update types

vvo committed:
feat(blob): add `clientPayload` and `tokenPayload` (#362)

* feat(blob): Allow passing clientPayload

This PR allows the use of `clientPayload` during `upload`, so you can attach
metadata to the file uploaded to Vercel Blob. This metadata will be sent as
`tokenPayload` by default in upload completed callbacks. And you can also enrich
it/configure it/verify it.

An example use-case of this feature would be to allow editing the images of a
blog post. To do so you would use `clientPayload: JSON.stringify({ postId: 345
})` in the browser. Then either in onBeforeGenerateToken or in onUploadCompleted
you would verify that the client making the call is allowed to modify this
postId (authorization) and update your database accordingly.

* update test

* update

* remove places where getTokenFromOptionsOrEnv is not needed

* add test without tokenPayload

vvo committed:
Merge branch 'main' into feat/add-clientPut

vvo committed:
upgrade API version given tokenPayload

vvo committed:
add changeset
